[feature] mentions

florian-sabonchi opened the pull request:
### Types
- [ ] Fix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Refactor (non-breaking change which improves code quality - QA thoroughly )
- [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (updates that would not affect or change the code involving the app itself)

### Changes

Added a dialog that allows to select users of a room to mark them in the chat. 

Attention! Known bug not all members of the room are currently loaded, unfortunately I can't find this bug. I have opened a pull request so that it is possible to review the code and possibly fix the bug.

Commit:
Fix extended profile widget

florian-sabonchi committed:
first draft for mention

florian-sabonchi committed:
first draft for mention

florian-sabonchi committed:
show mention added

florian-sabonchi committed:
show mention added

florian-sabonchi committed:
Avatar added for ListView

florian-sabonchi committed:
Avatar added for ListView

florian-sabonchi committed:
crossAxisAlignment added

florian-sabonchi committed:
crossAxisAlignment added

florian-sabonchi committed:
fixed overflow

florian-sabonchi committed:
fixed overflow

florian-sabonchi committed:
moved width to Mention

florian-sabonchi committed:
moved width to Mention

florian-sabonchi committed:
if a item user is clicked now the textField is updated

florian-sabonchi committed:
if a item user is clicked now the textField is updated

ereio committed:
Update styleguide.md

florian-sabonchi committed:
Cursor now moved to end

florian-sabonchi committed:
Cursor now moved to end

florian-sabonchi committed:
fixed visibility

florian-sabonchi committed:
fixed visibility

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -141,6 +148,33 @@ class ChatInputState extends State<ChatInput> {
+      mention = false;
+      users = props!.users;
+
+      if (text.startsWith('@')){
+        mention = true;
+        users = users.where((user) {
+          if (user != null){
+            final String searchText = text.toLowerCase();
+
+            if(user.userId != null){
+              return user.userId!.toLowerCase().contains(searchText);
+            }
+            else {
+              return user.displayName!.toLowerCase().contains(searchText);
+            }
+          }
+          else{
+            return false;
+          }
+        }).toList();
+      }
+    });
```
I suggest an `else` clause here, because otherwise you get an error when hitting backspace*:

![testingnull](https://user-images.githubusercontent.com/20861699/179395678-0219221a-25e9-49ee-ae59-235d56733ede.gif)

*in fact, you get the same error typing anything other than an `@` symbol

something like

```
if (text.startsWith('@')) {
    ...
}
else {
    users = [];
}
```

does the trick, although there may be a better way to address it

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -141,6 +148,33 @@ class ChatInputState extends State<ChatInput> {
+      mention = false;
+      users = props!.users;
+
+      if (text.startsWith('@')){
```
This means mentions only work at the beginning of the message, not in the body.

You could perhaps swap this to check `text.endsWith('@')`, as that would capture both usecases? Of course you'll need to change the internal logic to match

EdGeraghty reviewed (changes requested):
Your problem with the `roomUsers` is a problem with [the room Model](https://github.com/syphon-org/syphon/blob/dev/lib/store/rooms/room/model.dart). I note this is due for a refactor...

![image](https://user-images.githubusercontent.com/20861699/179396216-3113980b-3aaa-4bec-9bc4-530a0ea309b8.png)

Florian-Sabonchi commented:
> Your problem with the `roomUsers` is a problem with [the room Model](https://github.com/syphon-org/syphon/blob/dev/lib/store/rooms/room/model.dart). I note this is due for a refactor...
> 
> ![image](https://user-images.githubusercontent.com/20861699/179396216-3113980b-3aaa-4bec-9bc4-530a0ea309b8.png)

Does that mean I can ignore it for now?

florian-sabonchi committed:
fixed bug with mention on hitting backspace
for mor information see https://github.com/syphon-org/syphon/pull/706#discussion_r922816851

florian-sabonchi committed:
fixed bug with mention on hitting backspace
for mor information see https://github.com/syphon-org/syphon/pull/706#discussion_r922816851

florian-sabonchi committed:
mention now also possible in message body

florian-sabonchi committed:
mention now also possible in message body

florian-sabonchi committed:
Merge branch 'feature/mentions' of https://github.com/Florian-Sabonchi/syphon into feature/mentions

florian-sabonchi committed:
fixed reply position

EdGeraghty commented:
> Does that mean I can ignore it for now?

Yes, I believe so.

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -154,11 +154,12 @@ class ChatInputState extends State<ChatInput> {
       users = props!.users;
 
-      if (text.startsWith('@')){
+      if (text.startsWith('@') || text.contains(' @')) {
```
You can probably collapse this to just `if (text.contains('@')) {`

0x1a8510f2 replied:
That might pick up email addresses too which is probably not ideal

EdGeraghty replied:
Yes. Quite right. Derp!

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -172,6 +173,9 @@ class ChatInputState extends State<ChatInput> {
           }
         }).toList();
       }
+      else{
```
Can you please run the linter over your code? It's probably easiest to set it to run on save.

EdGeraghty reviewed (changes requested):
This is looking better, good work!

There is still a bug, though - I think you need to _append_ the userId instead of setting the textbox's text.

![testing-mentions](https://user-images.githubusercontent.com/20861699/181200201-e0db4506-014f-43cf-a712-ce31e8ede66b.gif)

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -373,7 +414,18 @@ class ChatInputState extends State<ChatInput> {
+              Container(padding: EdgeInsets.symmetric(horizontal: 15, vertical: 5),
+                  child: Mention(
+                    visible: showMention,
+                    width: MediaQuery.of(context).size.width - Dimensions.buttonSendSize * 2.1, // HACK: fix the width of the dialog
```
Where did this hack come from? I presume it's copied from somewhere else in the codebase...

Let's see what @ereio says when we finalise for his review.

EdGeraghty commented on lib/views/home/chat/widgets/chat-mention.dart:
```diff
@@ -0,0 +1,77 @@
+
+  onTab(User user){
+    setState((){
+      widget.controller.text = user.userId!;
```
Here we go - you want to be appending here instead of setting the text :)

florian-sabonchi committed:
fixed bug

florian-sabonchi committed:
fixed formatting

EdGeraghty commented on lib/views/home/chat/widgets/chat-mention.dart:
```diff
@@ -5,73 +5,67 @@ import 'package:syphon/store/user/model.dart';
-      widget.controller.text = user.userId!;
+  onTab(User user) {
+    setState(() {
+      widget.controller.text = '${widget.controller.text} ${user.userId}';
```
Getting very close, now!

![testing-mentions](https://user-images.githubusercontent.com/20861699/183285944-41cc7a90-e725-4557-b63c-3f61f3602040.gif)

florian-sabonchi committed:
fixed mention bug duplicate @

florian-sabonchi committed:
fixed bug

EdGeraghty commented:
You can take advantage of ` widget.controller.selection.baseOffset` to find out where the blinking cursor currently is in the input. From that, you should be able to support mentions anywhere in the text box :)

ereio commented on lib/views/home/chat/widgets/chat-mention.dart:
```diff
@@ -0,0 +1,75 @@
+  Widget build(BuildContext context) {
+    return ConstrainedBox(
+        constraints:
+            BoxConstraints(maxHeight: widget.height, maxWidth: widget.width),
```
this whole file, chat-mention.dart, looks pretty good

just make sure to follow the similar styling to the rest of the code base

in this instance, we should wrap the BoxConstraints params by adding a comma at the end.

this should auto format it to

```
BoxConstraints(
   maxHeight: widget.height, 
   maxWidth: widget.width,
)
```

this is so even with the line column width at 110, it will make it easier to read sequentially what parameters you've added and will require no additional wrapping when adding new potential styling

ereio commented:
Most of the requests I made above are styling changes. Happy to help make these - and again - find a way to automate some of this where you have no choice but to follow the convention. Dart / Flutter used to do a poor job at enforcement, where as other languages allow much more tooling around linting that makes it easier to just pull and auto format

I'll take a look this week

florian-sabonchi committed:
updated mention logic

florian-sabonchi committed:
updated setState

florian-sabonchi committed:
code Style updated

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -149,17 +148,26 @@ class ChatInputState extends State<ChatInput> {
             final String searchText = text
                 .toLowerCase()
-                .substring(text.indexOf('@') + 1, text.length);
+                .substring(text.lastIndexOf('@') + 1, text.length);
```
This probably needs to be relative to the current cursor position - this will just pick up the last one

Florian-Sabonchi replied:
Hello this works like this

EdGeraghty replied:
![testing-mentions](https://user-images.githubusercontent.com/20861699/188320022-596a4034-7487-4cc1-a54a-bfdcf6ee5708.gif)

I even went to the trouble of compiling, just to make sure.

EdGeraghty replied:
Another thing you can see in the above; there's a space character being inserted before the `@`

EdGeraghty replied:
![testing-mentions](https://user-images.githubusercontent.com/20861699/188320573-4d4278c7-6d44-4ca3-b8ec-96871e1016f8.gif)

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -149,17 +148,26 @@ class ChatInputState extends State<ChatInput> {
+    users = props!.users;
+
+      if(cursorPos > 0) {
+        setState((){
+          if(text[cursorPos - 1] == '@' && !mention) {
+            mention = true;
+          }
+          else if (text[cursorPos - 1] == ' ') {
+            mention = false;
+          }
+        });
```
Can you set the state further inside the `if`? As in, only if you actually need to toggle `mention`'s state?

EdGeraghty replied:
Perfect

EdGeraghty reviewed (changes requested):
So very close, now :)

florian-sabonchi committed:
updated state management

florian-sabonchi committed:
fixed bug

EdGeraghty commented on lib/views/home/chat/widgets/chat-input.dart:
```diff
@@ -149,17 +149,18 @@ class ChatInputState extends State<ChatInput> {
+    final int cursorPos = widget.controller.selection.baseOffset;
+    users = props!.users;
 
       if(cursorPos > 0) {
```
![image](https://user-images.githubusercontent.com/20861699/188320442-8e5c12e6-5165-4884-be11-1bff1e6bd8e4.png)

You also need to catch when there's no text but `mention == true`

EdGeraghty commented on lib/views/home/chat/widgets/chat-mention.dart:
```diff
@@ -65,9 +65,9 @@ class MentionState extends State<Mention> {
 
-      widget.controller.text = '${widget.controller.text.substring(0,
-          textLength - 1)} ${user.userId}'; // avoid duplicate @
+      widget.controller.text = '${text.substring(0, text.lastIndexOf("@"))} ${user.userId}';
```
- This is inserting a space at the beginning of what it's replacing
- The `lastIndexOf` will need addressing, similar to my comments on the mention searching

florian-sabonchi committed:
replaced mention logic with regex

florian-sabonchi committed:
code review

florian-sabonchi committed:
fixed bug

florian-sabonchi committed:
fixed code style

EdGeraghty commented:
I'm unable to test this

![image](https://user-images.githubusercontent.com/20861699/188725753-ea748031-97f6-426b-849f-9859fb5ca5f4.png)

florian-sabonchi committed:
fixed bug if user was null

EdGeraghty commented:
![image](https://user-images.githubusercontent.com/20861699/189499438-b0d3f941-df62-45e0-b5e7-ce79df03bdb7.png)

I invalidated all of my caches and built from scratch just to make sure

Florian-Sabonchi commented:
Perhaps something wrong with your settings

Commit:
add user null check

Commit:
linter

florian-sabonchi committed:
Merge pull request #3 from EdGeraghty/fix-null

User null fix

ereio commented:
I will pull this branch and clean it up before reviewing @Florian-Sabonchi. No need to make changes just yet

Aiming to have the 0.2.14 branch out next weekend and test the dev branch extensively throughout the week

ereio committed:
refactor: cleaned up mentions branch using chore/upgrades-and-cleanup, bugs addressed after refactor

ereio commented:
@Florian-Sabonchi pushed a branch here targeting your PR! I've refactored what's needed and will approve this PR after the merge. Great work! :)

https://github.com/Florian-Sabonchi/syphon/pull/4

florian-sabonchi committed:
Merge pull request #4 from syphon-org/refactor/mentions

refactor: cleaned up mentions branch

florian-sabonchi commented:
> @Florian-Sabonchi pushed a branch here targeting your PR! I've refactored what's needed and will approve this PR after the merge. Great work! :)
> 
> [Florian-Sabonchi#4](https://github.com/Florian-Sabonchi/syphon/pull/4)

Hi, Thanks for your work. I merged your pull request seems that we have some merge Konflikts because long time ago i made the initial Pull Request
